### PR TITLE
Give workloads a chance to gracefully shut down

### PR DESCRIPTION
**Pull request by autodidaddict**
Addresses #91 

Currently having trouble both generating a root file system and getting this new agent to perform a handshake. My local setup is so messed up I can't rule out environmental issues.

Would love someone else to try out this PR and make sure that we can start and stop workloads still.

### CODE REVIEW

**Comment by autodidaddict** [comment]
Terminal paste showing that you can now run devrun multiple times and the subscriptions clean up properly:
```
└─❯ ./nex/nex devrun ./examples/echoservice/echoservice nats_url=nats://192.168.127.1:4222
Reusing existing issuer account key: /home/derp/.nex/issuer.nk
Reusing existing publisher xkey: /home/derp/.nex/publisher.xk
Workload echoservice (cmp8idurih6p6mq86n3g) already exists on the target. Attempting to stop it
🚀 Workload 'echoservice' accepted. You can now refer to this workload with ID: cmp8jjmrih6p6mq86n40 on node NB6WTN573WUMO22RBYPLKNCOQ7ZS7KDIOTVCARSRPJ3SCQPLC6LY44M2%

└─❯ nats req svc.echo 'hey'
09:22:12 Sending request on "svc.echo"
09:22:12 Received with rtt 887.02µs
hey
```

File: examples/echoservice/main.go [hunk 1]
```diff
@@ -47,3 +51,24 @@ func main() {
...
+		signal.Reset(syscall.SIGINT, syscall.SIGTERM, syscall.SIGUSR1, syscall.SIGUSR2, syscall.SIGHUP)
+		c := make(chan os.Signal, 1)
+		signal.Notify(c, os.Interrupt, syscall.SIGTERM, syscall.SIGINT, syscall.SIGQUIT)
+
+		for {
+			switch s := <-c; {
+			case s == syscall.SIGTERM || s == os.Interrupt || s == syscall.SIGQUIT:
+				slog.Info("Caught signal, requesting clean shutdown", slog.String("signal", s.String()))
+				nc.Drain()
+				os.Exit(0)
```
Comment (jordan-rash) [inline examples/echoservice/main.go:66]: ~Is there any other option than `os.Exit`?~
Meh, i just noticed this is the echo example...should be fine

**Review by jordan-rash — DISMISSED** [review]

**Review by jordan-rash — DISMISSED** [review]

**Comment by JacobReynolds** [comment]
Trying this branch locally works on first deploy, but after that gives me `Failed to stop workload: nats: no responders available for request`. The first deploy also claims a timeout, but appears to work fine. The timeout happens with the release version as well. This could be pebcak, but wanted to provide another sample of data.

NATS server version: `Version:  2.10.9`

### nex cli output
```
root@nats-server:~/server# nats micro ls
nats: error: no micro instances found
root@nats-server:~/server# ../nex/nex/gracefulshutdown devrun example nats_url=nats://192.168.127.1:4222
Reusing existing issuer account key: /root/.nex/issuer.nk
Reusing existing publisher xkey: /root/.nex/publisher.xk
time=2024-01-25T17:38:31.991Z level=ERROR msg="failed to devrun workload" err="nats: timeout"
root@nats-server:~/server# nats micro ls
╭──────────────────────────────────────────────────────────────╮
│                      All Micro Services                      │
├─────────────┬─────────┬────────────────────────┬─────────────┤
│ Name        │ Version │ ID                     │ Description │
├─────────────┼─────────┼────────────────────────┼─────────────┤
│ EchoService │ 1.0.0   │ M3eK11kqiKeYQSsTun7HVl │             │
╰─────────────┴─────────┴────────────────────────┴─────────────╯

root@nats-server:~/server# nats request svc.echo hello
17:38:37 Sending request on "svc.echo"
17:38:37 Received with rtt 1.981777ms
Echo: hello


root@nats-server:~/server# ../nex/nex/gracefulshutdown devrun example nats_url=nats://192.168.127.1:4222
Reusing existing issuer account key: /root/.nex/issuer.nk
Reusing existing publisher xkey: /root/.nex/publisher.xk
Workload example (cmp9p053g9s8hme8g33g) already exists on the target. Attempting to stop it
time=2024-01-25T17:38:42.001Z level=ERROR msg="failed to devrun workload" err="Failed to stop workload: nats: no responders available for request"
root@nats-server:~/server# nats micro ls
╭──────────────────────────────────────────────────────────────╮
│                      All Micro Services                      │
├─────────────┬─────────┬────────────────────────┬─────────────┤
│ Name        │ Version │ ID                     │ Description │
├─────────────┼─────────┼────────────────────────┼─────────────┤
│ EchoService │ 1.0.0   │ M3eK11kqiKeYQSsTun7HVl │             │
╰─────────────┴─────────┴────────────────────────┴─────────────╯

root@nats-server:~/server# nats request svc.echo hello
17:38:46 Sending request on "svc.echo"
17:38:46 Received with rtt 1.53406ms
Echo: hello
```

### nex node output
```
root@nats-server:~/nex/nex# ./gracefulshutdown node up --config=../../nex.json 
time=2024-01-25T17:38:08.291Z level=INFO msg="Established node NATS connection" servers=nats://127.0.0.1:4222
time=2024-01-25T17:38:08.292Z level=INFO msg="Loaded node configuration from '%s'" config_path=../../nex.json
time=2024-01-25T17:38:08.296Z level=INFO msg="Virtual machine manager starting"
time=2024-01-25T17:38:08.378Z level=INFO msg="Internal NATs server started" client_url=nats://0.0.0.0:44631
time=2024-01-25T17:38:08.395Z level=INFO msg="Use this key as the recipient for encrypted run requests" public_xkey=XCEZHIAAKFGPV32M6E5JM6RFTEEAJW6SKK32NOXDCTAZUS7MTR6EZNL5
time=2024-01-25T17:38:08.396Z level=INFO msg="NATS execution engine awaiting commands" id=NDFFC5MKC7YPG235J3QXU3B5WICT34B3JUWFPXLXI7U4JADDCNWPFZEW version=0.0.1
time=2024-01-25T17:38:08.856Z level=INFO msg="Called startVMM(), setting up a VMM" firecracker=true vmmid=cmp9p053g9s8hme8g33g socket_path=/tmp/.firecracker.sock-34045-cmp9p053g9s8hme8g33g
time=2024-01-25T17:38:08.872Z level=INFO msg="VMM metrics disabled." firecracker=true vmmid=cmp9p053g9s8hme8g33g
time=2024-01-25T17:38:08.873Z level=INFO msg=refreshMachineConfiguration firecracker=true vmmid=cmp9p053g9s8hme8g33g err="[GET /machine-config][200] getMachineConfigurationOK  &{CPUTemplate: MemSizeMib:0xc00027d498 Smt:0xc00027d4a3 TrackDirtyPages:0xc00027d4a4 VcpuCount:0xc00027d490}"
time=2024-01-25T17:38:08.873Z level=INFO msg=PutGuestBootSource firecracker=true vmmid=cmp9p053g9s8hme8g33g err="[PUT /boot-source][204] putGuestBootSourceNoContent "
time=2024-01-25T17:38:08.873Z level=INFO msg="Attaching drive" firecracker=true vmmid=cmp9p053g9s8hme8g33g drive_path=/tmp/rootfs-cmp9p053g9s8hme8g33g.ext4 slot=1 root=true
time=2024-01-25T17:38:08.874Z level=INFO msg="Attached drive" firecracker=true vmmid=cmp9p053g9s8hme8g33g drive_path=/tmp/rootfs-cmp9p053g9s8hme8g33g.ext4 err="[PUT /drives/{drive_id}][204] putGuestDriveByIdNoContent "
time=2024-01-25T17:38:08.874Z level=INFO msg="Attaching NIC at index" firecracker=true vmmid=cmp9p053g9s8hme8g33g device_name=tap0 mac_addr=72:de:d9:08:93:6f interface_id=1
time=2024-01-25T17:38:08.958Z level=INFO msg="startInstance successful" firecracker=true vmmid=cmp9p053g9s8hme8g33g err="[PUT /actions][204] createSyncActionNoContent "
time=2024-01-25T17:38:08.966Z level=INFO msg="SetMetadata successful" firecracker=true vmmid=cmp9p053g9s8hme8g33g
time=2024-01-25T17:38:08.966Z level=INFO msg="Machine started" vmid=cmp9p053g9s8hme8g33g ip=192.168.127.61 gateway=192.168.127.1 netmask=ffffff00 hosttap=tap0 nats_host=192.168.127.1 nats_port=44631
time=2024-01-25T17:38:08.966Z level=INFO msg="Adding new VM to warm pool" ip=192.168.127.61 vmid=cmp9p053g9s8hme8g33g
time=2024-01-25T17:38:10.895Z level=INFO msg="Received agent handshake" vmid=cmp9p053g9s8hme8g33g message="Host-supplied metadata"
time=2024-01-25T17:38:31.244Z level=INFO msg="Attempting object store download" bucket=NEXCLIFILES key=example url=""
time=2024-01-25T17:38:31.667Z level=INFO msg="Successfully stored workload in internal object store" name=example
time=2024-01-25T17:38:31.667Z level=INFO msg="Submitting workload to VM" vmid=cmp9p053g9s8hme8g33g namespace=default workload=example type=elf
time=2024-01-25T17:38:32.298Z level=INFO msg="Called startVMM(), setting up a VMM" firecracker=true vmmid=cmp9p5t3g9s8hme8g340 socket_path=/tmp/.firecracker.sock-34045-cmp9p5t3g9s8hme8g340
time=2024-01-25T17:38:32.311Z level=INFO msg="VMM metrics disabled." firecracker=true vmmid=cmp9p5t3g9s8hme8g340
time=2024-01-25T17:38:32.311Z level=INFO msg=refreshMachineConfiguration firecracker=true vmmid=cmp9p5t3g9s8hme8g340 err="[GET /machine-config][200] getMachineConfigurationOK  &{CPUTemplate: MemSizeMib:0xc00a27f278 Smt:0xc00a27f283 TrackDirtyPages:0xc00a27f284 VcpuCount:0xc00a27f270}"
time=2024-01-25T17:38:32.312Z level=INFO msg=PutGuestBootSource firecracker=true vmmid=cmp9p5t3g9s8hme8g340 err="[PUT /boot-source][204] putGuestBootSourceNoContent "
time=2024-01-25T17:38:32.312Z level=INFO msg="Attaching drive" firecracker=true vmmid=cmp9p5t3g9s8hme8g340 drive_path=/tmp/rootfs-cmp9p5t3g9s8hme8g340.ext4 slot=1 root=true
time=2024-01-25T17:38:32.313Z level=INFO msg="Attached drive" firecracker=true vmmid=cmp9p5t3g9s8hme8g340 drive_path=/tmp/rootfs-cmp9p5t3g9s8hme8g340.ext4 err="[PUT /drives/{drive_id}][204] putGuestDriveByIdNoContent "
time=2024-01-25T17:38:32.313Z level=INFO msg="Attaching NIC at index" firecracker=true vmmid=cmp9p5t3g9s8hme8g340 device_name=tap0 mac_addr=66:4c:67:12:42:11 interface_id=1
time=2024-01-25T17:38:32.379Z level=INFO msg="Received agent event" vmid=cmp9p053g9s8hme8g33g type=workload_started
time=2024-01-25T17:38:32.379Z level=INFO msg="Work accepted" workload=example vmid=cmp9p053g9s8hme8g33g
time=2024-01-25T17:38:32.432Z level=INFO msg="startInstance successful" firecracker=true vmmid=cmp9p5t3g9s8hme8g340 err="[PUT /actions][204] createSyncActionNoContent "
time=2024-01-25T17:38:32.447Z level=INFO msg="SetMetadata successful" firecracker=true vmmid=cmp9p5t3g9s8hme8g340
time=2024-01-25T17:38:32.459Z level=INFO msg="Machine started" vmid=cmp9p5t3g9s8hme8g340 ip=192.168.127.62 gateway=192.168.127.1 netmask=ffffff00 hosttap=tap0 nats_host=192.168.127.1 nats_port=44631
time=2024-01-25T17:38:32.459Z level=INFO msg="Adding new VM to warm pool" ip=192.168.127.62 vmid=cmp9p5t3g9s8hme8g340
time=2024-01-25T17:38:34.547Z level=INFO msg="Received agent handshake" vmid=cmp9p5t3g9s8hme8g340 message="Host-supplied metadata"
time=2024-01-25T17:38:42.001Z level=ERROR msg="Failed to stop workload" err="nats: no responders available for request"
```

### nats server output
```
root@nats-server:~# nats-server -js
[34037] 2024/01/25 17:37:46.284935 [INF] Starting nats-server
[34037] 2024/01/25 17:37:46.285252 [INF]   Version:  2.10.9
[34037] 2024/01/25 17:37:46.285259 [INF]   Git:      [a5298e2]
[34037] 2024/01/25 17:37:46.285262 [INF]   Name:     NBBY5RNGH5PKOHOJENJQRBCN34TBIQBZCKELXH45V7JSIKMWXN2H2RXA
[34037] 2024/01/25 17:37:46.285269 [INF]   Node:     3YvDqjRf
[34037] 2024/01/25 17:37:46.285272 [INF]   ID:       NBBY5RNGH5PKOHOJENJQRBCN34TBIQBZCKELXH45V7JSIKMWXN2H2RXA
[34037] 2024/01/25 17:37:46.285811 [INF] Starting JetStream
[34037] 2024/01/25 17:37:46.286011 [INF]     _ ___ _____ ___ _____ ___ ___   _   __  __
[34037] 2024/01/25 17:37:46.286027 [INF]  _ | | __|_   _/ __|_   _| _ \ __| /_\ |  \/  |
[34037] 2024/01/25 17:37:46.286030 [INF] | || | _|  | | \__ \ | | |   / _| / _ \| |\/| |
[34037] 2024/01/25 17:37:46.286033 [INF]  \__/|___| |_| |___/ |_| |_|_\___/_/ \_\_|  |_|
[34037] 2024/01/25 17:37:46.286035 [INF] 
[34037] 2024/01/25 17:37:46.286038 [INF]          https://docs.nats.io/jetstream
[34037] 2024/01/25 17:37:46.286041 [INF] 
[34037] 2024/01/25 17:37:46.286043 [INF] ---------------- JETSTREAM ----------------
[34037] 2024/01/25 17:37:46.286048 [INF]   Max Memory:      2.86 GB
[34037] 2024/01/25 17:37:46.286099 [INF]   Max Storage:     80.81 GB
[34037] 2024/01/25 17:37:46.286112 [INF]   Store Directory: "/tmp/nats/jetstream"
[34037] 2024/01/25 17:37:46.286123 [INF] -------------------------------------------
[34037] 2024/01/25 17:37:46.286878 [INF]   Starting restore for stream '$G > OBJ_NEXCLIFILES'
[34037] 2024/01/25 17:37:46.287275 [INF]   Restored 66 messages for stream '$G > OBJ_NEXCLIFILES' in 0s
[34037] 2024/01/25 17:37:46.287704 [INF] Listening for client connections on 0.0.0.0:4222
[34037] 2024/01/25 17:37:46.292270 [INF] Server is ready
```

**Comment by jordan-rash** [comment]
Thanks @JacobReynolds.  Per our Slack convo, the published rootfs has a wrong permission right now.  If you build it locally, it should be fixed.  We will cut a release later today with an updated rootfs

**Review by jordan-rash — DISMISSED** [review]
LGTM

**Review by jordan-rash — APPROVED** [review]

**Comment by autodidaddict** [comment]
no idea what's going wrong

**Comment by autodidaddict** [comment]
please work

**Review by jordan-rash — APPROVED** [review]

**Comment by burakakca** [comment]
Hi,

```bash
nats-server
[128423] 2024/03/01 16:16:19.229992 [INF] Starting nats-server
[128423] 2024/03/01 16:16:19.230314 [INF]   Version:  2.10.11
[128423] 2024/03/01 16:16:19.230323 [INF]   Git:      [5fc6051]
[128423] 2024/03/01 16:16:19.230329 [INF]   Name:     NA3UPF2KFNGQWU7KKHG4ESWKGIO2XGZYVYOHMXC3LKX4Y4A37ECR7E2V
[128423] 2024/03/01 16:16:19.230349 [INF]   ID:       NA3UPF2KFNGQWU7KKHG4ESWKGIO2XGZYVYOHMXC3LKX4Y4A37ECR7E2V
[128423] 2024/03/01 16:16:19.232238 [INF] Listening for client connections on 0.0.0.0:4222
[128423] 2024/03/01 16:16:19.232916 [INF] Server is ready

sudo nex node preflight --config=./simple.json
[sudo] password for user: 
Validating - Required CNI Plugins
	  🔎Searching - /opt/cni/bin 
	  ✅ Dependency Satisfied - /opt/cni/bin/host-local [host-local CNI plugin]
	  ✅ Dependency Satisfied - /opt/cni/bin/ptp [ptp CNI plugin]
	  ✅ Dependency Satisfied - /opt/cni/bin/tc-redirect-tap [tc-redirect-tap CNI plugin]

Validating - Required binaries
	  🔎Searching - /usr/local/bin 
	  ✅ Dependency Satisfied - /usr/local/bin/firecracker [Firecracker VM binary]

Validating - CNI configuration requirements
	  🔎Searching - /etc/cni/conf.d 
	  ✅ Dependency Satisfied - /etc/cni/conf.d/fcnet.conflist [CNI Configuration]

Validating - VMLinux Kernel
	  ✅ Dependency Satisfied - /path/to/vmlinux-5.10 [VMLinux Kernel]

Validating - Root Filesystem Template
	  ✅ Dependency Satisfied - /path/to/rootfs.ext4 [Root Filesystem Template]

user@userbn:~/Desktop/nats$ sudo nex node up --config=./simple.json
time=2024-03-01T16:36:30.686+03:00 level=INFO msg="Generated keypair for node" public_key=NDDCGXKZ5Z3P43KDTMNPLXZYBCCL7REOXYZ6GMQKBWHWQDLNCZSMMXRC
time=2024-03-01T16:36:30.686+03:00 level=INFO msg="Loaded node configuration" config_path=./simple.json
time=2024-03-01T16:36:30.686+03:00 level=INFO msg="Initialized telemetry"
time=2024-03-01T16:36:30.689+03:00 level=INFO msg="Established node NATS connection" servers=""
time=2024-03-01T16:36:30.702+03:00 level=INFO msg="Internal NATS server started" client_url=nats://0.0.0.0:38009
time=2024-03-01T16:36:30.702+03:00 level=INFO msg="Use this key as the recipient for encrypted run requests" public_xkey=XDEHXWJMQ6ZCLFRFKSWLX2BQ7MLCUCYZGEQAPAJL7VMZUZK2JPICVKCT
time=2024-03-01T16:36:30.702+03:00 level=INFO msg="NATS execution engine awaiting commands" id=NDDCGXKZ5Z3P43KDTMNPLXZYBCCL7REOXYZ6GMQKBWHWQDLNCZSMMXRC version=0.1.4
time=2024-03-01T16:36:30.702+03:00 level=INFO msg="Virtual machine manager starting"
time=2024-03-01T16:36:30.702+03:00 level=INFO msg="Resetting network"
time=2024-03-01T16:36:30.703+03:00 level=INFO msg="Publishing node started event"
time=2024-03-01T16:36:31.194+03:00 level=INFO msg="Called startVMM(), setting up a VMM" firecracker=true vmmid=cngtjngpd7rvbdfklck0 socket_path=/tmp/.firecracker.sock-129425-cngtjngpd7rvbdfklck0
time=2024-03-01T16:36:31.208+03:00 level=INFO msg="VMM metrics disabled." firecracker=true vmmid=cngtjngpd7rvbdfklck0
time=2024-03-01T16:36:31.209+03:00 level=INFO msg=refreshMachineConfiguration firecracker=true vmmid=cngtjngpd7rvbdfklck0 err="[GET /machine-config][200] getMachineConfigurationOK  &{CPUTemplate: MemSizeMib:0xc0003574f8 Smt:0xc000357503 TrackDirtyPages:0xc000357504 VcpuCount:0xc0003574f0}"
time=2024-03-01T16:36:31.210+03:00 level=INFO msg=PutGuestBootSource firecracker=true vmmid=cngtjngpd7rvbdfklck0 err="[PUT /boot-source][204] putGuestBootSourceNoContent "
time=2024-03-01T16:36:31.210+03:00 level=INFO msg="Attaching drive" firecracker=true vmmid=cngtjngpd7rvbdfklck0 drive_path=/tmp/rootfs-cngtjngpd7rvbdfklck0.ext4 slot=1 root=true
time=2024-03-01T16:36:31.211+03:00 level=INFO msg="Attached drive" firecracker=true vmmid=cngtjngpd7rvbdfklck0 drive_path=/tmp/rootfs-cngtjngpd7rvbdfklck0.ext4 err="[PUT /drives/{drive_id}][204] putGuestDriveByIdNoContent "
time=2024-03-01T16:36:31.211+03:00 level=INFO msg="Attaching NIC at index" firecracker=true vmmid=cngtjngpd7rvbdfklck0 device_name=tap0 mac_addr=ea:a3:32:2f:7b:4e interface_id=1
time=2024-03-01T16:36:31.369+03:00 level=INFO msg="startInstance successful" firecracker=true vmmid=cngtjngpd7rvbdfklck0 err="[PUT /actions][204] createSyncActionNoContent "
time=2024-03-01T16:36:31.369+03:00 level=INFO msg="Machine started" vmid=cngtjngpd7rvbdfklck0 ip=192.168.127.2 gateway=192.168.127.1 netmask=ffffff00 hosttap=tap0 nats_host=192.168.127.1 nats_port=38009
time=2024-03-01T16:36:31.370+03:00 level=INFO msg="SetMetadata successful" firecracker=true vmmid=cngtjngpd7rvbdfklck0
time=2024-03-01T16:36:31.370+03:00 level=INFO msg="Adding new VM to warm pool" ip=192.168.127.2 vmid=cngtjngpd7rvbdfklck0
time=2024-03-01T16:36:33.246+03:00 level=INFO msg="Received agent handshake" vmid=cngtjngpd7rvbdfklck0 message="Host-supplied metadata"


nex node info NDDCGXKZ5Z3P43KDTMNPLXZYBCCL7REOXYZ6GMQKBWHWQDLNCZSMMXRC
NEX Node Information

       Node: NDDCGXKZ5Z3P43KDTMNPLXZYBCCL7REOXYZ6GMQKBWHWQDLNCZSMMXRC
       Xkey: XDEHXWJMQ6ZCLFRFKSWLX2BQ7MLCUCYZGEQAPAJL7VMZUZK2JPICVKCT
    Version: 0.1.4
     Uptime: 16m18s
       Tags: nex.arch=amd64, nex.cpucount=16, nex.os=linux, simple=true

Memory in kB:

         Free: 435,036
    Available: 12,001,120
        Total: 16,365,564
```

**Problem**:

```bash
nex devrun ./echoservice/echoservice nats_url="nats://192.168.127.1:4222"
Reusing existing issuer account key: /home/user/.nex/issuer.nk
Reusing existing publisher xkey: /home/user/.nex/publisher.xk
time=2024-03-01T16:57:20.172+03:00 level=ERROR msg="failed to devrun workload" err="nats: no responders available for request"

```

**Pull request merged**